ccl/kvccl/kvfollowerreadsccl: TestBoundedStalenessDataDriven failed

ccl/kvccl/kvfollowerreadsccl.TestBoundedStalenessDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20550525?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20550525?buildTab=artifacts#/) on master @ [ff839799faaf1e9ebda21e09e959e47f697d821a](https://github.com/cockroachdb/cockroach/commits/ff839799faaf1e9ebda21e09e959e47f697d821a):


```
=== RUN   TestBoundedStalenessDataDriven
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/8c04cb0c40c192808857b5e756290a7c/logTestBoundedStalenessDataDriven3294506660
    test_log_scope.go:82: use -show-logs to present logs inline
    boundedstaleness_test.go:434: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8c04cb0c40c192808857b5e756290a7c/logTestBoundedStalenessDataDriven3294506660
--- FAIL: TestBoundedStalenessDataDriven (7.69s)
=== RUN   TestBoundedStalenessDataDriven/single_row
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/14970/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:5:
          exec-system-tenant
          SET CLUSTER SETTING kv.closed_timestamp.target_duration = '1s';
          ----
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/14970/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:9:
          exec
          CREATE TABLE t(pk INT PRIMARY KEY) WITH (schema_locked = false);
          INSERT INTO t VALUES (1);
          ----
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/14970/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:18:
          query idx=2 wait-until-match
          SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('1μs') WHERE pk = 1
          ----
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/14970/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:25:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_min_timestamp(now() - '1μs') WHERE pk = 1
        
        expected:
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
        
        found:
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read
--- FAIL: TestBoundedStalenessDataDriven/single_row (7.67s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBoundedStalenessDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-55086

ccl/kvccl/kvfollowerreadsccl.TestBoundedStalenessDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20555924?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20555924?buildTab=artifacts#/) on master @ [e1d612210b299df836bdbdb95b5e6c979529a813](https://github.com/cockroachdb/cockroach/commits/e1d612210b299df836bdbdb95b5e6c979529a813):


```
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:141:
          exec-system-tenant
          SET CLUSTER SETTING kv.closed_timestamp.target_duration = '1hr';
          ----
    datadriven.go:352: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:145: still running after 10.004855s
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:145:
          exec
          ALTER TABLE t ADD COLUMN new_col INT NOT NULL DEFAULT 2
          ----
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:151:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('10s') WHERE pk = 1
          ----
          1 2
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:158:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_min_timestamp(now() - '10s') WHERE pk = 1
          ----
          1 2
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:165:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('10s', false) WHERE pk = 1
          ----
          1 2
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:172:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_min_timestamp(now() - '10s', false) WHERE pk = 1
          ----
          1 2
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    datadriven.go:352: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:184: still running after 10.003845128s
    datadriven.go:352: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:184: still running after 20.004143791s
    datadriven.go:352: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:184: still running after 30.005023647s
    datadriven.go:352: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17294/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:184: still running after 40.008759084s
    boundedstaleness_test.go:394: condition failed to evaluate within 45s: from boundedstaleness_test.go:419: not follower reads found:
        events (0 found):
--- FAIL: TestBoundedStalenessDataDriven/single_row (64.40s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBoundedStalenessDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/kvccl/kvfollowerreadsccl.TestBoundedStalenessDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20570529?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20570529?buildTab=artifacts#/) on master @ [881d2d7ef6bbd88beeb44636e1d1812c41b803aa](https://github.com/cockroachdb/cockroach/commits/881d2d7ef6bbd88beeb44636e1d1812c41b803aa):


```
=== RUN   TestBoundedStalenessDataDriven
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/8c04cb0c40c192808857b5e756290a7c/logTestBoundedStalenessDataDriven165195559
    test_log_scope.go:82: use -show-logs to present logs inline
    boundedstaleness_test.go:434: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8c04cb0c40c192808857b5e756290a7c/logTestBoundedStalenessDataDriven165195559
--- FAIL: TestBoundedStalenessDataDriven (9.21s)
=== RUN   TestBoundedStalenessDataDriven/single_row
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11646/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:5:
          exec-system-tenant
          SET CLUSTER SETTING kv.closed_timestamp.target_duration = '1s';
          ----
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11646/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:9:
          exec
          CREATE TABLE t(pk INT PRIMARY KEY) WITH (schema_locked = false);
          INSERT INTO t VALUES (1);
          ----
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11646/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:18:
          query idx=2 wait-until-match
          SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('1μs') WHERE pk = 1
          ----
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11646/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:25:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_min_timestamp(now() - '1μs') WHERE pk = 1
          ----
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11646/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:32:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('1μs', false) WHERE pk = 1
        
        expected:
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
        
        found:
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read
--- FAIL: TestBoundedStalenessDataDriven/single_row (9.15s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBoundedStalenessDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/kvccl/kvfollowerreadsccl.TestBoundedStalenessDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20628917?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20628917?buildTab=artifacts#/) on master @ [2593003795fcb177bd0caa1d23dd40fe856b2e28](https://github.com/cockroachdb/cockroach/commits/2593003795fcb177bd0caa1d23dd40fe856b2e28):


```
=== RUN   TestBoundedStalenessDataDriven
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/18f40673c89ae073fd9b4688a4790099/logTestBoundedStalenessDataDriven832908380
    test_log_scope.go:82: use -show-logs to present logs inline
    boundedstaleness_test.go:434: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/18f40673c89ae073fd9b4688a4790099/logTestBoundedStalenessDataDriven832908380
--- FAIL: TestBoundedStalenessDataDriven (8.34s)
=== RUN   TestBoundedStalenessDataDriven/single_row
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/14061/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:5:
          exec-system-tenant
          SET CLUSTER SETTING kv.closed_timestamp.target_duration = '1s';
          ----
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/14061/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:9:
          exec
          CREATE TABLE t(pk INT PRIMARY KEY) WITH (schema_locked = false);
          INSERT INTO t VALUES (1);
          ----
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/14061/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:18:
          query idx=2 wait-until-match
          SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('1μs') WHERE pk = 1
          ----
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/14061/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:25:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_min_timestamp(now() - '1μs') WHERE pk = 1
        
        expected:
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
        
        found:
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read
--- FAIL: TestBoundedStalenessDataDriven/single_row (8.31s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBoundedStalenessDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/kvccl/kvfollowerreadsccl.TestBoundedStalenessDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20636749?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20636749?buildTab=artifacts#/) on master @ [b59054be548797afcd4a8154e37dabb2d8aa0e61](https://github.com/cockroachdb/cockroach/commits/b59054be548797afcd4a8154e37dabb2d8aa0e61):


```
=== RUN   TestBoundedStalenessDataDriven
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/18f40673c89ae073fd9b4688a4790099/logTestBoundedStalenessDataDriven2518719795
    test_log_scope.go:82: use -show-logs to present logs inline
    boundedstaleness_test.go:434: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/18f40673c89ae073fd9b4688a4790099/logTestBoundedStalenessDataDriven2518719795
--- FAIL: TestBoundedStalenessDataDriven (8.67s)
=== RUN   TestBoundedStalenessDataDriven/single_row
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17668/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:5:
          exec-system-tenant
          SET CLUSTER SETTING kv.closed_timestamp.target_duration = '1s';
          ----
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17668/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:9:
          exec
          CREATE TABLE t(pk INT PRIMARY KEY) WITH (schema_locked = false);
          INSERT INTO t VALUES (1);
          ----
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17668/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:18:
          query idx=2 wait-until-match
          SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('1μs') WHERE pk = 1
          ----
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17668/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:25:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_min_timestamp(now() - '1μs') WHERE pk = 1
          ----
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
    boundedstaleness_test.go:315: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17668/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:32:
          query idx=2
          SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('1μs', false) WHERE pk = 1
        
        expected:
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read then remote leaseholder read
        
        found:
          1
          events (1 found):
           * event 1: colbatchscan trace on node_idx 2: local read
--- FAIL: TestBoundedStalenessDataDriven/single_row (8.63s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBoundedStalenessDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/kvccl/kvfollowerreadsccl.TestBoundedStalenessDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20741262?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20741262?buildTab=artifacts#/) on master @ [a97e76d739b75fa4e86c3302c1a925c4a16a63ed](https://github.com/cockroachdb/cockroach/commits/a97e76d739b75fa4e86c3302c1a925c4a16a63ed):


```
=== RUN   TestBoundedStalenessDataDriven
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/8c04cb0c40c192808857b5e756290a7c/logTestBoundedStalenessDataDriven3214120961
    test_log_scope.go:82: use -show-logs to present logs inline
    boundedstaleness_test.go:445: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8c04cb0c40c192808857b5e756290a7c/logTestBoundedStalenessDataDriven3214120961
--- FAIL: TestBoundedStalenessDataDriven (52.96s)
=== RUN   TestBoundedStalenessDataDriven/single_row
    boundedstaleness_test.go:320: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11936/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:5:
          exec-system-tenant
          SET CLUSTER SETTING kv.closed_timestamp.target_duration = '1s';
          ----
    boundedstaleness_test.go:320: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11936/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:9:
          exec
          CREATE TABLE t(pk INT PRIMARY KEY) WITH (schema_locked = false);
          INSERT INTO t VALUES (1);
          ----
    boundedstaleness_test.go:320: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11936/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:15:
          query ignore-events
          SELECT DISTINCT(lease_holder) FROM [SHOW RANGES FROM TABLE t WITH DETAILS];
          ----
          1
    datadriven.go:357: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11936/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:24: still running after 10.000889738s
    datadriven.go:357: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11936/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:24: still running after 20.001707372s
    datadriven.go:357: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11936/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:24: still running after 30.002127873s
    datadriven.go:357: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11936/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:24: still running after 40.002414732s
    boundedstaleness_test.go:405: condition failed to evaluate within 45s: from boundedstaleness_test.go:436: not yet a match, output:
        1
        events (1 found):
         * event 1: colbatchscan trace on node_idx 2: local read
    datadriven.go:343: 
        fatal during /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11936/execroot/com_github_cockroachdb_cockroach/bazel-out/s390x-fastbuild/bin/pkg/ccl/kvccl/kvfollowerreadsccl/kvfollowerreadsccl_test_/kvfollowerreadsccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row:24:
        SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('1μs') WHERE pk = 1
--- FAIL: TestBoundedStalenessDataDriven/single_row (52.94s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBoundedStalenessDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/kvccl/kvfollowerreadsccl/boundedstaleness_test.go
+++ b/pkg/ccl/kvccl/kvfollowerreadsccl/boundedstaleness_test.go
@@ -45,6 +45,10 @@ var (
 	)
 )
 
+// fullTraceDebug is a flag that controls whether full traces are printed in the
+// case of some errors.
+const fullTraceDebug = false
+
 func TestBoundedStalenessEnterpriseLicense(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
@@ -148,8 +152,10 @@ type boundedStalenessEvents struct {
 	// A mutex is needed as the event handlers (onStmtTrace) can race.
 	mu struct {
 		syncutil.Mutex
-		stmt   string
-		events []boundedStalenessDataDrivenEvent
+		stmt string
+		// Only populated if fullTraceDebug constant is true.
+		traceForDebugging string
+		events            []boundedStalenessDataDrivenEvent
 	}
 }
 
@@ -165,6 +171,13 @@ func (bse *boundedStalenessEvents) clearEvents() {
 	bse.mu.Lock()
 	defer bse.mu.Unlock()
 	bse.mu.events = nil
+}
+
+func (bse *boundedStalenessEvents) fullTrace() string {
+	bse.mu.Lock()
+	defer bse.mu.Unlock()
+
+	return bse.mu.traceForDebugging
 }
 
 func (bse *boundedStalenessEvents) setStmt(s string) {
@@ -240,17 +253,23 @@ func (bse *boundedStalenessEvents) onStmtTrace(nodeIdx int, rec tracingpb.Record
 	defer bse.mu.Unlock()
 
 	if bse.mu.stmt != "" && bse.mu.stmt == stmt {
+		if fullTraceDebug {
+			bse.mu.traceForDebugging = rec.String()
+		}
+
 		spans := make(map[tracingpb.SpanID]tracingpb.RecordedSpan)
 		for _, sp := range rec {
 			spans[sp.SpanID] = sp
+			notLeaseHolderError := tracing.LogsContainMsg(sp, "[NotLeaseHolderError] lease held by different store;")
+			notLeaseHolderError = notLeaseHolderError || tracing.LogsContainMsg(sp, "[NotLeaseHolderError] leader lease is not held locally, cannot determine validity;")
+
 			if sp.Operation == "dist sender send" && spans[sp.ParentSpanID].Operation == "colbatchscan" {
 				bse.mu.events = append(bse.mu.events, &boundedStalenessTraceEvent{
-					operation:    spans[sp.ParentSpanID].Operation,
-					nodeIdx:      nodeIdx,
-					localRead:    tracing.LogsContainMsg(sp, kvbase.RoutingRequestLocallyMsg),
-					followerRead: kvtestutils.OnlyFollowerReads(rec),
-					remoteLeaseholderRead: tracing.LogsContainMsg(sp, "[NotLeaseHolderError] lease held by different store;") &&
-						tracing.LogsContainMsg(sp, "trying next peer"),
+					operation:             spans[sp.ParentSpanID].Operation,
+					nodeIdx:               nodeIdx,
+					localRead:             tracing.LogsContainMsg(sp, kvbase.RoutingRequestLocallyMsg),
+					followerRead:          kvtestutils.OnlyFollowerReads(rec),
+					remoteLeaseholderRead: notLeaseHolderError && tracing.LogsContainMsg(sp, "trying next peer"),
 				})
 			}
 		}
@@ -261,10 +280,7 @@ func TestBoundedStalenessDataDriven(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	const msg = "1μs staleness reads may actually succeed due to the slow environment"
-	skip.UnderStress(t, msg)
-	skip.UnderRace(t, msg)
-	skip.UnderDeadlock(t, msg)
+	skip.UnderDuress(t, "1μs staleness reads may actually succeed due to the slow environment")
 	defer ccl.TestingEnableEnterprise()()
 
 	ctx := context.Background()
@@ -426,14 +442,22 @@ func TestBoundedStalenessDataDriven(t *testing.T) {
 						}
 					}()
 					if !followerRead {
+						var trace string
+						if fullTraceDebug {
+							trace = fmt.Sprintf("\nfull_trace:\n%s", bse.fullTrace())
+						}
 						bse.clearEvents()
-						return errors.AssertionFailedf("not follower reads found:\n%s", bse.String())
+						return errors.AssertionFailedf("not follower reads found:\n%s%s", bse.String(), trace)
 					}
 				}
 				if waitUntilMatch {
 					if d.Expected != ret {
+						var trace string
+						if fullTraceDebug {
+							trace = fmt.Sprintf("\nfull_trace:\n%s", bse.fullTrace())
+						}
 						bse.clearEvents()
-						return errors.AssertionFailedf("not yet a match, output:\n%s\n", ret)
+						return errors.AssertionFailedf("not yet a match, output:\n%s%s", ret, trace)
 					}
 				}
 				return nil

